pkg/operator/controller/instance_controller_test.go

```diff
@@ -0,0 +1,319 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller_test
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"testing"
+	"time"
+
+	. "github.com/onsi/gomega"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+
+	clusterlink "github.com/clusterlink-net/clusterlink/pkg/apis/clusterlink.net/v1alpha1"
+	"github.com/clusterlink-net/clusterlink/pkg/operator/controller"
+)
+
+const (
+	ClusterLinkName = "cl-job"
+	Timeout         = time.Second * 30
+	duration        = time.Second * 10
+	IntervalTime    = time.Millisecond * 250
+)
+
+var (
+	k8sClient client.Client
+	ctx       context.Context
+	cancel    context.CancelFunc
+)
+
+func TestMain(m *testing.M) {
+	// Bootstrap test environment
+	ctx, cancel = context.WithCancel(context.TODO())
+	testEnv := &envtest.Environment{
+		CRDDirectoryPaths:     []string{filepath.Join("..", "..", "..", "config", "operator", "crds")},
+		ErrorIfCRDPathMissing: true,
+		BinaryAssetsDirectory: filepath.Join("..", "..", "..", "bin", "k8s",
+			fmt.Sprintf("1.28.3-%s-%s", runtime.GOOS, runtime.GOARCH)),
+	}
+
+	var err error
+	cfg, err := testEnv.Start()
+	if err != nil {
+		fmt.Printf("Failed to start test environment: %v\n", err)
+		cancel()
+		return
+	}
+
+	err = clusterlink.AddToScheme(scheme.Scheme)
+	if err != nil {
+		fmt.Printf("Failed to add API to scheme: %v\n", err)
+		cancel()
+		return
+	}
+
+	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme.Scheme})
+	if err != nil {
+		fmt.Printf("Failed to create Kubernetes client: %v\n", err)
+		cancel()
+		return
+	}
+
+	k8sManager, err := ctrl.NewManager(cfg, ctrl.Options{
+		Scheme: scheme.Scheme,
+	})
+	if err != nil {
+		fmt.Printf("Failed to create controller manager: %v\n", err)
+		cancel()
+		return
+	}
+
+	err = (&controller.InstanceReconciler{
+		Client:        k8sManager.GetClient(),
+		Scheme:        k8sManager.GetScheme(),
+		Logger:        logrus.WithField("component", "reconciler"),
+		InstancesMeta: make(map[string]string),
+	}).SetupWithManager(k8sManager)
+	if err != nil {
+		fmt.Printf("Failed to set up reconciler: %v\n", err)
+		cancel()
+		return
+	}
+
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				fmt.Printf("Recovered from panic: %v\n", r)
+			}
+		}()
+		err := k8sManager.Start(ctx)
+		if err != nil {
+			fmt.Printf("Failed to run controller manager: %v\n", err)
+		}
+	}()
+
+	code := m.Run()
+
+	// Cancelling the context
+	cancel()
```

Comment:
if `m.Run` is synchronous, whatis the purpose of calling `cancel`? Context cancelling is usually done to interrupt a blocking go routine.

Reply:
I used it for cleanup the context